src/coins.cpp

```diff
@@ -260,6 +295,230 @@ void CCoinsViewCache::BringBestAnchorIntoCache(
+        h = h - 1;
+
+        // drafting left child
+        draftMMRNode(entry_indices, entries, GetHistoryAt(epochId, left_pos-1), h, left_pos);
+
+        // drafting right child
+        draftMMRNode(entry_indices, entries, GetHistoryAt(epochId, right_pos-1), h, right_pos);
+
+        // continuing on right slope
+        peak_pos = right_pos;
+    }    
+
+    return total_peaks;
+}
+
+HistoryCache& CCoinsViewCache::SelectHistoryCache(uint32_t epochId) const {
+    auto entry = historyCacheMap.find(epochId);
+
+    if (entry == historyCacheMap.end()) {
+        historyCacheMap[epochId] = HistoryCache(
+            base->GetHistoryLength(epochId),
+            base->GetHistoryRoot(epochId),
+            epochId
+        );
+
+        return historyCacheMap.find(epochId)->second;
+    }   
+
+    return entry->second;
+}
+
+void CCoinsViewCache::PushHistoryNode(uint32_t epochId, const HistoryNode node) {
+    auto treeLength = GetHistoryLength(epochId);
+    HistoryCache& historyCache = SelectHistoryCache(epochId);
+    
+    if (treeLength == 0) {
+        // special case, it just goes into the cache right away
+        historyCache.Extend(node);
+
+        librustzcash_mmr_hash_node(epochId, node.data(), historyCache.root.begin());
+
+        return;
+    }
+
+    std::vector<HistoryEntry> entries;
+    std::vector<uint32_t> entry_indices;
+
+    PreloadHistoryTree(epochId, false, entries, entry_indices);
+
+    uint256 newRoot;
+    std::array<HistoryNode, 32> appendBuf;
+
+    uint32_t appends = librustzcash_mmr_append(
+        epochId, 
+        treeLength,
+        entry_indices.data(), 
+        entries.data()->data(),
+        entry_indices.size(),
+        node.data(),
+        newRoot.begin(),
+        appendBuf.data()->data()
+    );
+    
+    for (size_t i = 0; i < appends; i++) {
+        historyCache.Extend(appendBuf[i]);
+    }
+
+    historyCache.root = newRoot;
+}
+
+void CCoinsViewCache::PopHistoryNode(uint32_t epochId) {
+    auto treeLength = GetHistoryLength(epochId);
+    HistoryCache& historyCache = SelectHistoryCache(epochId);
+
+    // elementary pops near the head of tree
+    if (treeLength == 0) {
+        // Caller is not expected to pop from empty tree! Caller should
+        // switch to previous epoch and pop history from there.
+        throw std::runtime_error("popping hisory node from empty history");
+    } else if (treeLength == 1) {
+        // Just resetting tree to empty
+        historyCache.Truncate(0);
+        historyCache.root = uint256();
+        return;
+    } else if (treeLength == 3) {
+        historyCache.Truncate(1);
+        uint256 newRoot;
+        if (librustzcash_mmr_hash_node(
+            epochId, 
+            libzcash::LeafToEntry(GetHistoryAt(epochId, 0)).data(), 
+            newRoot.begin()
+        ) == 0) {
+            historyCache.root = newRoot;
+        } else {
+            throw std::runtime_error("hasing node failed");
+        }
+    }
```

Comment:
In a pairing, @daira and I determined that the `return` was missing, and confirmed that `treelength == 2` cannot occur (because `treeLength` refers to the length of the MMR stored-node array structure, not the number of leaves).